common/src/main/java/io/druid/math/expr/Evals.java

```diff
@@ -40,4 +46,39 @@ public static Number toNumber(Object value)
     }
     return longValue;
   }
+
+  public static boolean isConstant(Expr expr)
+  {
+    return expr instanceof DoubleExpr || expr instanceof LongExpr || expr instanceof StringExpr;
+  }
+
+  public static boolean isAllConstants(Expr... exprs)
+  {
+    return isAllConstants(Arrays.asList(exprs));
+  }
+
+  public static boolean isAllConstants(List<Expr> exprs)
+  {
+    for (Expr expr : exprs) {
+      if (!isConstant(expr)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // for binary operator not providing constructor of form <init>(String, Expr, Expr),
+  // you should create it explicitly in here
+  public static Expr binaryOp(BinaryOpExprBase binary, Expr left, Expr right)
+  {
+    try {
+      return binary.getClass()
+                   .getDeclaredConstructor(String.class, Expr.class, Expr.class)
+                   .newInstance(binary.op, left, right);
+    }
+    catch (Exception e) {
+      log.info(e, "failed to rewrite expression " + binary);
```

Comment:
Do we expect this to happen "normally" or is this really a problem indicating a bug? If it's going to happen "normally" it shouldn't be `log.info` as that will be too noisy in the logs. If it indicates a bug, please bump it up to `log.warn`.